src/charm.py

```diff
@@ -239,75 +217,24 @@ def _cookie_encryption_key(self) -> Optional[str]:
-        # Define container configuration
-        container = {
-            "override": "replace",
-            "summary": "identity platform login ui",
-            "command": self._login_ui_service_command,
-            "startup": "enabled",
-            "environment": {
-                "HYDRA_ADMIN_URL": self._get_hydra_endpoint_info(),
-                "KRATOS_PUBLIC_URL": kratos_info.get("public_endpoint", ""),
-                "KRATOS_ADMIN_URL": kratos_info.get("admin_endpoint", ""),
-                "PORT": str(APPLICATION_PORT),
-                "BASE_URL": self._domain_url,
-                "COOKIES_ENCRYPTION_KEY": self._cookie_encryption_key,
-                "TRACING_ENABLED": False,
-                "AUTHORIZATION_ENABLED": False,
-                "LOG_LEVEL": self._log_level,
-                "DEBUG": self._log_level == "DEBUG",
-            },
-        }
-
-        if self._kratos_info.is_ready():
-            container["environment"]["MFA_ENABLED"] = literal_eval(kratos_info.get("mfa_enabled"))
-            container["environment"]["OIDC_WEBAUTHN_SEQUENCING_ENABLED"] = literal_eval(
-                kratos_info.get("oidc_webauthn_sequencing_enabled")
-            )
+        return self._pebble_service.render_pebble_layer(
+            self._domain_url,
+            self._cookie_encryption_key,
+            self._log_level,
+            HydraEndpointData.load(self.hydra_endpoints),
+            KratosInfoData.load(self._kratos_info),
+            TracingData.load(self.tracing),
+        )
 
-        if self._tracing_ready:
-            container["environment"]["OTEL_HTTP_ENDPOINT"] = self.tracing.get_endpoint("otlp_http")
-            container["environment"]["OTEL_GRPC_ENDPOINT"] = self.tracing.get_endpoint("otlp_grpc")
-            container["environment"]["TRACING_ENABLED"] = True
-
-        # Define Pebble layer configuration
-        pebble_layer = {
-            "summary": "login_ui layer",
-            "description": "pebble config layer for identity platform login ui",
-            "services": {WORKLOAD_CONTAINER_NAME: container},
-            "checks": {
-                "login-ui-alive": {
-                    "override": "replace",
-                    "http": {"url": f"http://localhost:{APPLICATION_PORT}/api/v0/status"},
-                },
-            },
-        }
-        return Layer(pebble_layer)
-
-    def _get_kratos_info(self) -> Dict:
-        kratos_info = {}
-        if self._kratos_info.is_ready():
-            try:
-                kratos_info = self._kratos_info.get_kratos_info()
-            except KratosInfoRelationDataMissingError:
-                logger.info("No kratos-info relation data found")
-        return kratos_info
-
-    def _update_login_ui_endpoint_relation_data(self, event: RelationEvent) -> None:
+    def _update_login_ui_endpoint_relation_data(self, _: RelationEvent) -> None:
         endpoint = self._domain_url or ""
 
         self.endpoints_provider.send_endpoints_relation_data(
```

Comment:
Not sure about this: Should we wrap this logic in a class in `integrations.py`?

Reply:
So I've seen other charms do the "main" relation logic in the charm py, I'd leave it as is but I don't have strong opinions here